go/arrow/ipc/reader.go

```diff
@@ -196,7 +196,9 @@ func (r *Reader) Read() (array.Record, error) {
 	}
 
 	if !r.next() {
-		if r.done {
+		if r.err != nil {
+			return nil, r.err
+		} else if r.done {
 			return nil, io.EOF
 		}
 		return nil, r.err
```

Comment:
I think this could be cleaner:

```go
if r.done && r.err != nil {
        return nil, io.EOF
}
return nil, r.err
```

Reply:
The problem is `next()` sets both `done` and `err`, so this will still return `io.EOF` instead of the actual error.

https://github.com/apache/arrow/blob/c9495abbd0ad36575714808c563599c498969087/go/arrow/ipc/reader.go#L163-L167

Reply:
Maybe the change should be to instead not set `done` when we get an error?

Reply:
How would my example return `io.EOF` in the case of an error? By doing `if r.done && r.err == nil` we only return the `io.EOF` if there isn't an error to return.

Reply:
Sorry, I thought I saw `err != nil` for some reason. I'll update it.